keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -99,182 +103,173 @@ File and block are first class citizens within the Kubernetes ecosystem.  Object
-1. The user creates an OBC in the app’s namespace which references the pre-existing OB.
-1. The central controller detects the OBC and notices its `objectBucketRef` is filled in but its `secretRef` is empty.
-1. The central controller updates OB Phase and OBC reference.
-1. The Sidecar detects the OB update and calls the `Grant()` gRPC method.
-1. The Driver grants access to the bucket and returns pertinent endpoint, credential, and metadata.
-1. The Sidecar creates a secret in its namespace containing essential connection information.
-1. The central controller clones the secret to the OBC’s namespace, binds the OB, and the app is ready to run.
-
-### Revoke Bucket Access (needs to handle stratic and dynamic brownfield!)
-
-Reminder: BucketClass is ignored in browfield operations.
-
-1. The user deletes the OBC, which blocks until the finalizer is removed.
-1. The central controller detects the OBC change and updates OB’s Phase.
-1. The Sidecar detects the OB update and calls the `Revoke()` gRPC method.
-1. The Driver revokes access to the bucket and responds with no errors.
-1. The Sidecar deletes the OB’s secret.
-1. The central controller deletes the secret’s copy in the app’s namespace.
-1. The central controller removes the finalizers from the OBC, OB, and Secret.
-1. The OBC and its secret are garbage collected and the OB remains.
+### Greenfield Buckets
+
+#### Create Bucket
+
+1. The user creates Bucket in the app’s namespace.
+1. The COSI Controller detects the Bucket and creates a BucketContent object containing driver-relevant data.
+1. The Sidecar detects the BucketContent and issues a CreateBucket() rpc with any parameters defined in the BucketClass.
+1. The driver provisions the bucket and returns pertinent endpoint, credential, and metadata information.
+1. The sidecar creates a Secret in its namespace (`cosi-system`) containing endpoint and credential information and an owener reference to the BucketContent.
+1. rThe COSI controller generates a Secret in the Bucket's namespace containing the data of it's parent Secret, with an owner reference to the Bucket.  
+1. The workflow ingests the Secret to begin operation.
+
+#### Delete Bucket
+
+1. The user deletes thier Bucket, which blocks until backend deletion operations complete.
+1. The COSI controller detects the update and deletes the BucketContent, which also blocks until backend deletion operations complete.
+1. The sidecar detects this and issues a DeleteBucket() rpc to the Driver.  It passes pertitent data stored in the BucketContent.
+1. The driver deletes the bucket from the object store and responds with no error.
+1. The sidecar removes the BucketContent's finalizer.  The BucketContent and parent Secret are garbage collected.
+1. The COSI controller removes the finalizer on the Bucket.  The Bucket and the child Secret are garbage collected.
+
+### Brownfield Buckets
+
+#### Grant Access
+
+1. A BucketClass is defined specifically referencing an object store bucket.
```

Comment:
Thus, a 1-1 mapping of brownfield buckets and bucket classes. Will there be a scaling concern here if we need to support many brownfield buckets?

Reply:
I don't think scale should be a concern for N number of buckets exposed to a cluster.  The scale comes from how many individual connections need to be made per bucket.  Exposing them as bucket classes allows us to automate the connection (Bucket:BucketContent), which I think is where the real scaling challenges lie.

Reply:
We had a similar approach in the bucket lib where the storage class named the brownfield bucket. A criticism of that design was the potential proliferation of many, very similar storage classes with the only/main difference being the name of the existing bucket. I see the same issue here. However, if we prefer that the `bucket` resource not name the brownfield bucket, then I don't yet have a better plan...

Reply:
Using bucket classes is the most scalable solution in this case.  You either have to create 1 BC per object store bucket, or 'n' BucketContents per connection, per bucket.  For the admin's workload, it's O(n) vs O(n*m), where n = object store buckets and m = connections to a bucket.  If he has to create just classes, the O(n*m) is automated away by the COSI controller.